File: src/uam/sdk/webhook_verify.py
"""Receiver-side webhook signature verification (HOOK-03).

Small utility for webhook receivers to verify that incoming webhook
payloads were signed by the UAM relay.  Uses HMAC-SHA256 with the
agent's token as the shared secret.

Usage::

    from uam.sdk.webhook_verify import verify_webhook_signature

    payload = request.body()  # raw bytes
    signature = request.headers["X-UAM-Signature"]
    token = "your-agent-token"

    if verify_webhook_signature(payload, signature, token):
        # payload is authentic
        ...
"""

from __future__ import annotations

import hashlib
import hmac


def verify_webhook_signature(
    payload_bytes: bytes,
    signature_header: str,
    token: str,
) -> bool:
    """Verify an HMAC-SHA256 webhook signature.

    Args:
        payload_bytes: The raw request body bytes.
        signature_header: The ``X-UAM-Signature`` header value
            (format: ``sha256=<hex>``).
        token: The agent's token (shared secret).

    Returns:
        ``True`` if the signature is valid, ``False`` otherwise.
        Uses ``hmac.compare_digest`` for constant-time comparison
        to prevent timing attacks.
    """
    if not signature_header.startswith("sha256="):
        return False

    expected = hmac.new(
        token.encode("utf-8"),
        payload_bytes,
        hashlib.sha256,
    ).hexdigest()
    received = signature_header[len("sha256="):]
    return hmac.compare_digest(expected, received)

File: ts-sdk/src/sdk/webhook-verify.ts
/**
 * Receiver-side webhook signature verification (HOOK-03).
 *
 * Small utility for webhook receivers to verify that incoming webhook
 * payloads were signed by the UAM relay.  Uses HMAC-SHA256 with the
 * agent's token as the shared secret.
 *
 * Matches Python webhook_verify.py API surface.
 *
 * Usage:
 *
 *     import { verifyWebhookSignature } from "uam";
 *
 *     const payload = Buffer.from(requestBody);
 *     const signature = request.headers["x-uam-signature"];
 *     const token = "your-agent-token";
 *
 *     if (verifyWebhookSignature(payload, signature, token)) {
 *       // payload is authentic
 *     }
 */

import { createHmac, timingSafeEqual } from "node:crypto";

/**
 * Verify an HMAC-SHA256 webhook signature.
 *
 * @param payload - Raw request body bytes
 * @param signatureHeader - "sha256=<hex>" format (X-UAM-Signature header)
 * @param token - Shared secret (agent token)
 * @returns `true` if valid, `false` otherwise.
 *
 * Uses `crypto.timingSafeEqual()` for constant-time comparison
 * to prevent timing attacks.
 */
export function verifyWebhookSignature(
  payload: Buffer | Uint8Array,
  signatureHeader: string,
  token: string
): boolean {
  if (!signatureHeader.startsWith("sha256=")) {
    return false;
  }

  const expected = createHmac("sha256", token)
    .update(payload)
    .digest("hex");
  const received = signatureHeader.slice("sha256=".length);

  // Constant-time comparison requires equal-length buffers
  const expectedBuf = Buffer.from(expected, "utf-8");
  const receivedBuf = Buffer.from(received, "utf-8");

  if (expectedBuf.length !== receivedBuf.length) {
    return false;
  }

  return timingSafeEqual(expectedBuf, receivedBuf);
}
